src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsAdornmentManager.cs

```diff
@@ -0,0 +1,237 @@
+                    {
+                        continue;
+                    }
+
+                    var lineNum = mappedPoint.Value.GetContainingLine().LineNumber;
+                    if (!map.TryGetValue((lineNum, changedSpan), out var list))
+                    {
+                        list = new List<IMappingTagSpan<InlineDiagnosticsTag>>();
+                        map.Add((lineNum, changedSpan), list);
+                    }
+
+                    list.Add(tagMappingSpan);
+                }
+            }
+
+            return map;
+        }
+
+        protected override void UpdateSpans_CallOnlyOnUIThread(NormalizedSnapshotSpanCollection changedSpanCollection, bool removeOldTags)
+        {
+            Contract.ThrowIfNull(changedSpanCollection);
+
+            // this method should only run on UI thread as we do WPF here.
+            Contract.ThrowIfFalse(TextView.VisualElement.Dispatcher.CheckAccess());
+
+            var viewLines = TextView.TextViewLines;
+            if (viewLines == null || viewLines.Count == 0)
+            {
+                return; // nothing to draw on
+            }
+
+            // removing is a separate pass from adding so that new stuff is not removed.
+            if (removeOldTags)
+            {
+                foreach (var changedSpan in changedSpanCollection)
+                {
+                    // is there any effect on the view?
+                    if (viewLines.IntersectsBufferSpan(changedSpan))
+                    {
+                        AdornmentLayer.RemoveAdornmentsByVisualSpan(changedSpan);
+                    }
+                }
+            }
+
+            var map = GetSpansOnEachLine(changedSpanCollection);
+            var tagSpanToPointMap = GetTagSpansToSnapshotPointMap(map);
+            foreach (var tagMappingSpanList in map.Values)
+            {
+                var tagMappingSpan = GetHighestOrderTag(tagMappingSpanList);
+                TryMapToSingleSnapshotSpan(tagMappingSpan.Span, TextView.TextSnapshot, out var span);
+                var geometry = viewLines.GetMarkerGeometry(span);
+                if (geometry != null)
+                {
+                    var tag = tagMappingSpan.Tag;
+                    var classificationType = _classificationRegistryService.GetClassificationType(InlineDiagnosticsTag.TagID + tag.ErrorType);
+                    var graphicsResult = tag.GetGraphics(TextView, geometry, GetFormat(classificationType));
+                    if (!tagSpanToPointMap.TryGetValue(tagMappingSpan, out var point))
+                    {
+                        continue;
+                    }
+
+                    var lineView = TextView.GetTextViewLineContainingBufferPosition(point);
+
+                    var visualElement = graphicsResult.VisualElement;
+                    if (tag.Location is InlineDiagnosticsLocations.PlacedAtEndOfCode)
+                    {
+                        Canvas.SetLeft(visualElement, lineView.Right);
```

Comment:
if the line doesn't end with any spacing, do we at least give a small amount of buffer to prevent things from looking too cramped?

Reply:
there is a margin on the tag itself giving space on the left. I can either keep that margin on the tag or place it further left on the screen.